### PR TITLE
DEX-822 Devnet deployment via Jenkins

### PR DESCRIPTION
**Pull request by ngadiyak**

### CODE REVIEW

File: deploy.Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,97 @@
+def deployNode (host) {
+    return {
+         sshagent (credentials: ['buildagent-matcher']) {
+             sh "ssh -o StrictHostKeyChecking=no -l buildagent-matcher ${host} hostname"
+
+             sh "scp ./waves/node/target/waves-devnet*all*.deb buildagent-matcher@${host}:/home/buildagent-matcher"
+             sh "scp ./matcher/dex-load/src/main/resources/reinstallNode.sh buildagent-matcher@${host}:/home/buildagent-matcher"
+             sleep time: 80000, unit: 'MILLISECONDS'
```
Comment (vsuharnikov) [inline deploy.Jenkinsfile:8]: Why?

File: deploy.Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        }
+        stage('Compile Matcher') {
+            steps {
+                dir('matcher') {
+                    sh 'printenv'
+                    sh 'git fetch --tags'
+                    sh 'find ~/.sbt/1.0/staging/*/waves -type d -name target | xargs -I{} rm -rf {}'
+                    sh 'find . -type d -name target | xargs -I{} rm -rf {}'
+                    sh 'sbt "set Global / scalacOptions ++= Seq(\\"-Xfatal-warnings\\", \\"-Ywarn-unused:-imports\\");session save;cleanAll;compile;packageAll"'
```
Comment (vsuharnikov) [inline deploy.Jenkinsfile:44]: We don't need here warn checks, so just:
```
sh 'sbt packageAll'
```

File: deploy.Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                    sh "scp ./matcher/dex-load/src/main/resources/reinstallExtension.sh buildagent-matcher@${NODE4}:/home/buildagent-matcher"
+                    sh "scp ./matcher/dex-load/src/main/resources/reinstallMatcher.sh buildagent-matcher@${MATCHER}:/home/buildagent-matcher"
+                    sleep time: 80000, unit: 'MILLISECONDS'
+                    sh "ssh -q buildagent-matcher@${NODE4} sudo sh reinstallExtension.sh"
+                    sh "ssh -q buildagent-matcher@${MATCHER} sudo sh reinstallMatcher.sh"
+                }
+            }
+        }
+    }
+}
```
Comment (vsuharnikov) [inline deploy.Jenkinsfile:97]: A newline is required

File: deploy.Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        stage('Deploy Matcher') {
+            steps {
+                sshagent (credentials: ['buildagent-matcher']) {
+                    sh "ssh -o StrictHostKeyChecking=no -l buildagent-matcher ${MATCHER} hostname"
+                    sh "scp ./matcher/target/release/waves-dex*all*.deb buildagent-matcher@${MATCHER}:/home/buildagent-matcher"
+                    sh "scp ./matcher/target/release/waves-dex-extension-devnet*all*.deb buildagent-matcher@${NODE4}:/home/buildagent-matcher"
+                    sh "scp ./matcher/dex-load/src/main/resources/reinstallExtension.sh buildagent-matcher@${NODE4}:/home/buildagent-matcher"
+                    sh "scp ./matcher/dex-load/src/main/resources/reinstallMatcher.sh buildagent-matcher@${MATCHER}:/home/buildagent-matcher"
+                    sleep time: 80000, unit: 'MILLISECONDS'
```
Comment (vsuharnikov) [inline deploy.Jenkinsfile:90]: Also here

File: dex-load/src/main/resources/reinstallExtension.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+systemctl stop waves-devnet.service || true
+dpkg -r waves-dex-extension-devnet || true
+dpkg -P waves-dex-extension-devnet || true
```
Comment (vsuharnikov) [inline dex-load/src/main/resources/reinstallExtension.sh:3]: `dpkg -P waves-dex-extension-devnet || true` is sufficient, see https://askubuntu.com/a/128381

File: dex-load/src/main/resources/reinstallMatcher.sh [hunk 6]
```diff
@@ -0,0 +1,8 @@
+systemctl stop waves-dex.service || true
+rm -rf /var/lib/waves-dex/data || true
+dpkg -r waves-dex || true
+dpkg -P waves-dex || true
```
Comment (vsuharnikov) [inline dex-load/src/main/resources/reinstallMatcher.sh:4]: Same

File: dex-load/src/main/resources/reinstallExtension.sh [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+dpkg -i /home/buildagent-matcher/waves-dex-extension-devnet*.deb
```
Comment (vsuharnikov) [inline dex-load/src/main/resources/reinstallExtension.sh:4]: I didn't see cleanup after previous deployments in build steps. So this line probably try to install multiple versions of waves-dex-extension-devnet

File: deploy.Jenkinsfile (hunk 1 above)
Comment (ngadiyak) [reply deploy.Jenkinsfile:8]: unfortunately, jenkins doesn't wait for scp command finished... so we need to wait for sure that shell script will be success

File: deploy.Jenkinsfile (hunk 2 above)
Comment (ngadiyak) [reply deploy.Jenkinsfile:44]: ok

File: deploy.Jenkinsfile (hunk 4 above)
Comment (ngadiyak) [reply deploy.Jenkinsfile:90]: I've gotten the answer upper... may me I try to find a better way to exclude sleep step

File: dex-load/src/main/resources/reinstallExtension.sh (hunk 5 above)
Comment (ngadiyak) [reply dex-load/src/main/resources/reinstallExtension.sh:3]: OK

File: deploy.Jenkinsfile [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        SBT_THREAD_NUMBER = "${SBT_THREAD_NUMBER}"
+        SBT_OPTS = '-Xmx2g -XX:ReservedCodeCacheSize=128m -XX:+CMSClassUnloadingEnabled -Dnetwork=devnet'
+        PATH = "${env.SBT_HOME}/bin:${env.PATH}"
+    }
+    stages {
+        stage('Checkout') {
+            steps {
+                cleanWs()
+                dir('matcher') {
+                    checkout scm: [$class: 'GitSCM', userRemoteConfigs: [[url: 'https://github.com/wavesplatform/matcher.git', credentialsId: null]], extensions: [[$class: 'CloneOption', noTags: false, reference: '', shallow: false]], branches: [[name: '*/DEX-822-devnet-deployment-via-jenkins']]], poll: false
```
Comment (vsuharnikov) [inline deploy.Jenkinsfile:28]: `DEX-822-devnet-deployment-via-jenkins` ?

**Review by vsuharnikov — APPROVED** [review]

**Pull request merged**